modules/stochastic_tools/src/utils/GaussianProcess.C

```diff
@@ -0,0 +1,371 @@
+                                     const RealEigenMatrix & training_data,
+                                     const GPOptimizerOptions & opts)
+{
+  std::vector<Real> theta(_num_tunable, 0.0);
+  _covariance_function->buildHyperParamMap(_hyperparam_map, _hyperparam_vec_map);
+
+  mapToVec(_tuning_data, _hyperparam_map, _hyperparam_vec_map, theta);
+
+  // Internal params for Adam; set to the recommended values in the paper
+  Real b1 = opts.b1;
+  Real b2 = opts.b2;
+  Real eps = opts.eps;
+
+  std::vector<Real> m0(_num_tunable, 0.0);
+  std::vector<Real> v0(_num_tunable, 0.0);
+
+  Real new_val;
+  Real m_hat;
+  Real v_hat;
+  Real store_loss = 0.0;
+  std::vector<Real> grad1;
+
+  // Initialize randomizer
+  std::vector<unsigned int> v_sequence(training_params.rows());
+  std::iota(std::begin(v_sequence), std::end(v_sequence), 0);
+  RealEigenMatrix inputs(_batch_size, training_params.cols());
+  RealEigenMatrix outputs(_batch_size, training_data.cols());
+  if (opts.show_optimization_details)
+    Moose::out << "OPTIMIZING GP HYPER-PARAMETERS USING Adam" << std::endl;
+  for (unsigned int ss = 0; ss < opts.num_iter; ++ss)
+  {
+    // Shuffle data
+    MooseRandom generator;
+    generator.seed(0, 1980);
+    generator.saveState();
+    MooseUtils::shuffle<unsigned int>(v_sequence, generator, 0);
+    for (unsigned int ii = 0; ii < _batch_size; ++ii)
+    {
+      for (unsigned int jj = 0; jj < training_params.cols(); ++jj)
+        inputs(ii, jj) = training_params(v_sequence[ii], jj);
+
+      for (unsigned int jj = 0; jj < training_data.cols(); ++jj)
+        outputs(ii, jj) = training_data(v_sequence[ii], jj);
+    }
+
+    store_loss = getLoss(inputs, outputs);
+    if (opts.show_optimization_details)
+      Moose::out << "Iteration: " << ss + 1 << " LOSS: " << store_loss << std::endl;
+    grad1 = getGradient(inputs);
+    for (auto iter = _tuning_data.begin(); iter != _tuning_data.end(); ++iter)
+    {
+      const auto first_index = std::get<0>(iter->second);
+      const auto num_entries = std::get<1>(iter->second);
+      for (unsigned int ii = 0; ii < num_entries; ++ii)
+      {
+        const auto global_index = first_index + ii;
+        m0[global_index] = b1 * m0[global_index] + (1 - b1) * grad1[global_index];
+        v0[global_index] =
+            b2 * v0[global_index] + (1 - b2) * grad1[global_index] * grad1[global_index];
+        m_hat = m0[global_index] / (1 - std::pow(b1, (ss + 1)));
+        v_hat = v0[global_index] / (1 - std::pow(b2, (ss + 1)));
+        new_val = theta[global_index] - opts.learning_rate * m_hat / (std::sqrt(v_hat) + eps);
+
+        const auto min_value = std::get<2>(iter->second);
+        const auto max_value = std::get<3>(iter->second);
+
+        theta[global_index] = std::min(std::max(new_val, min_value), max_value);
+      }
+    }
+    vecToMap(_tuning_data, _hyperparam_map, _hyperparam_vec_map, theta);
+    _covariance_function->loadHyperParamMap(_hyperparam_map, _hyperparam_vec_map);
```

Comment:
Here we may add an if statement like `if (opts.show_optimization_details && ss % show_every_n_iteration == 0)` then show the loss value. `show_every_n_iteration` is another GP optimization parameter to be added as an option.

Reply:
Added